src/Utils.php

```diff
@@ -137,28 +137,46 @@ public static function subscriptionData(\PaymentLineItem $line_item) {
    *
    * @param \PaymentLineItem $line_item
    *   A recurring line item.
+   * @param \DateTimeImmutable $now
+   *   The date and time considered to be now, defaults to now in UTC.
    *
    * @return \DateTime|null
    *   The calculated start date or
    *   `null` if the line item doesn’t include any date settings.
    */
-  public static function getStartDate(\PaymentLineItem $line_item) {
+  public static function getStartDate(\PaymentLineItem $line_item, \DateTimeImmutable $now = NULL) {
     $recurrence = $line_item->recurrence;
     if (empty($recurrence->start_date) && empty($recurrence->month) && empty($recurrence->day_of_month)) {
       return NULL;
     }
-    // Earliest possible start date.
-    $earliest = $recurrence->start_date ?? new \DateTime('tomorrow', new \DateTimeZone('UTC'));
-    // Date meeting day of month and month requirements.
-    $y = $earliest->format('Y');
+    $now = $now ?? new \DateTimeImmutable('', new \DateTimeZone('UTC'));
+    // Earliest possible date, either tomorrow or future recurrence start date.
+    $earliest = $now->modify('+1 day');
+    if (!empty($recurrence->start_date) && $recurrence->start_date > $earliest) {
+      $earliest = $recurrence->start_date;
+    }
+    // Date in the past meeting day of month and month requirements.
+    $y = $earliest->format('Y') - 1;
     $m = $recurrence->month ?? $earliest->format('m');
     $d = $recurrence->day_of_month ?? $earliest->format('d');
-    $date = new \DateTime($y . $m . $d, new \DateTimeZone('UTC'));
+    $date = $now->setDate($y, $m, $d);
     // Find the first matching date after the earliest.
-    $unit = rtrim($recurrence->interval_unit, 'ly');
-    $count = $recurrence->interval_value ?? 1;
+    $unit = 'month';
+    $count = 1;
+    if (!empty($recurrence->month)) {
+      $interval_unit = $recurrence->interval_unit;
+      if ($interval_unit == 'yearly') {
+        $unit = 'year';
+      }
+      elseif ($interval_unit == 'monthly' && !empty($interval_value = $recurrence->interval_value)) {
+        $count = 12 % $interval_value == 0 ? $interval_value : $count;
```

Comment:
It would be better to explicitly have `1` as else-value. We don’t want to keep the value (whatever it is) we want to set it to `1`.

Reply:
Re-wrote the whole if statement, there was another glitch in it too (weekly payments without month). I hope it’s clearer now!